src/eradiate/scenes/surface/_dem.py

```diff
@@ -1,61 +1,204 @@
 from __future__ import annotations
 
+import os.path
 import typing as t
 import warnings
+from pathlib import Path
 
 import attrs
+import mitsuba as mi
 import numpy as np
 import pint
 import xarray as xr
-from pinttr.util import ensure_units
 
 from ._core import Surface
-from ..bsdfs import BSDF, LambertianBSDF, bsdf_factory
-from ..core import InstanceSceneElement, NodeSceneElement
-from ..shapes import BufferMeshShape, FileMeshShape, Shape, shape_factory
+from ..bsdfs import BSDF, LambertianBSDF, OpacityMaskBSDF, bsdf_factory
+from ..core import traverse
+from ..geometry import PlaneParallelGeometry, SphericalShellGeometry
+from ..shapes import (
+    BufferMeshShape,
+    FileMeshShape,
+    RectangleShape,
+    Shape,
+    SphereShape,
+    shape_factory,
+)
 from ...attrs import documented, get_doc, parse_docs
+from ...exceptions import OverriddenValueWarning
+from ...kernel import TypeIdLookupStrategy
 from ...units import to_quantity
 from ...units import unit_context_config as ucc
 from ...units import unit_registry as ureg
 
 
-@parse_docs
-@attrs.define(eq=False, slots=False)
-class DEMSurface(Surface):
+def _generate_dem_vertices(latitude_points, longitude_points, elevation):
     """
-    DEM surface [``dem``]
+    Generate DEM vertex positions as (latitude, longitude, elevation) pairs
+    for a plane parallel atmosphere geometry.
+
+    Parameters
+    ----------
+    geometry: .SceneGeometry
+        Atmospheric geometry
+
+    Returns
+    -------
+    vertices: list
+        DEM vertices as (lat, lon, elevation) pairs. Atmospheric geometry
+        is not taken into account at this step
+    """
+    vertices = []
+    lat_range = list(latitude_points)
+    lon_range = list(longitude_points)
+
+    for i, theta in enumerate(lat_range):
+        for j, phi in enumerate(lon_range):
+            z = elevation[i, j].m_as(ucc.get("length"))
+            vertices.append([theta, phi, z])
+    return vertices
```

Comment:
This function has a few issues.

* The loop-based implementation is wasteful (and the in-loop Pint conversion adds even more penalty). The `z` array is just [`elevation.ravel()`](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html) (maybe you'll have to transpose, I can't remember exactly) and the `vertices` array is then `theta.ravel()`, `phi.ravel()` and `elevation.ravel()` stacked. Such a formulation will be vectorized automatically by Numpy.
* The returned value has no units. This is fine for internal implementation details, but at least then please mention the mandatory conversion in the docs. And it's then best to convert to fixed units, not context-dependent ones. Good news though: if you vectorize this, applying Pint units will be cheap.
* Outdated docstring.

Reply:
The values inside each tuple have different units. I can't just attach one unit to the entire list, because it would not be correct.